Assignments/Program.cs

```diff
@@ -1 +1,131 @@
+             thirdDigit = 3 (3rd digit of part)
+
+      3. The state machine starts with the state of HUNDREDS as the maximum place value of any 3-digit number is 100.
+     
+      Overall, the first loop segregates the input string into three 3-digit numbers and supplies to the second loop,
+      which converts the supplied 3-digit number to equivalent words.  Finally all the words are added together and returned.
+       */
+
+      // FIRST for-loop:
+      for (int i = 0; i < 3; i++) {
+         int part = int.Parse (input.Substring (i * 3, 3));
+         State s = State.Hundreds;
+         // SECOND for-loop:
+         for (int n = 3; n > 0; n--) {
+            int nthDigit = part / (int)Math.Pow (10, n - 1) % 10, thirdDigit = part % 10;
+            switch (s, nthDigit) {
+               case (State.Hundreds, >= 0 and <= 9): {
+                     s = State.Tens;
+                     if (nthDigit != 0) {
+                        output += ones[nthDigit] + " Hundred ";
+                        if (part / 10 % 10 > 0 || thirdDigit > 0) output += "and ";
+                     }
+                     break;
+                  }
+               case (State.Tens, 1): {
+                     s = State.End;
+                     output += ones[10 + thirdDigit] + " ";
+                     break;
+                  }
+               case (State.Tens, 0 or (>= 2 and <= 9)): {
+                     s = State.Ones;
+                     if (nthDigit != 0) output += tens[nthDigit] + ((i == 2 && thirdDigit > 0) ? "-" : " ");
+                     break;
+                  }
+               case (State.Ones, >= 0 and <= 9): {
+                     s = State.End;
+                     if (nthDigit != 0) output += ones[nthDigit] + " ";
+                     break;
+                  }
+               default: { s = State.End; break; }
+            }
+         }
+         if (part > 0 && i != 2) {
+            int lastThreeDigit = num / 1000;
+            var (cond, key) = (i == 0) ? (num / 1000 % 1000 > 0 || lastThreeDigit > 0, "Million") : (lastThreeDigit > 0, "Thousand");
+            output += key + (cond ? ", " : " ");
+         }
+      }
+      return output;
+   }
+   enum State { Hundreds, Tens, Ones, End };
+
+   /// <summary>Converts input number to Roman numerals</summary>
+   public static StringBuilder ToRomans (int num) {
+      StringBuilder output = new ();
+      if (num == 0) output.Append ("Zero has no symbol in Roman numerals system.");
+      else {
+         int[] value = { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1 };
+         string[] roman = { "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I" };
+         int v = Array.Find (value, a => a <= num);
+         for (int i = Array.IndexOf (value, v); i < value.Length; i++) {
+            int x = value[i], y = num / x;
+            output.Append (string.Concat (Enumerable.Repeat (roman[i], y)));

```

Comment:
This still has to be refined. If I need the Roman value of 5 this loop is unnecessarily iterating thrice and in the first iteration appends V and then in other iterations nothing needs to be appended but it iterates. Check this.